### PR TITLE
send multiple TTC in response to multi-WTCT

### PR DESCRIPTION
**Pull request by tworec**
Multiple TTC are sent as a response to `num_subtasks` filed in WTCT.
Resolves (partly) golemfactory/golem-unlimited#216

The general idea is to transit from a __one subtask__ per single Provider to **one WTCT** per single Provider. It is done by remembering WTCT msg hash, then counting started and accepted subtasks.

Limitation for just one WTCT is hardcoded and not parametrized intentionally. Current Provider code is not amended, so WTCT will have `num_subtasks == 1` for standard Golem Brass Providers.  Golem Unlimited is supposed to use this functionality solely.

The Provider is allowed to ask for multiple TTC expressing it via `num_subtasks` parameter in WTCT. If Provider's offer will be selected he will get at most `num_subtasks` TTC's. It can be less when there is no more work to perform. 

Until all assigned work (TTC's) are accepted by Requestor, Provider is not able to ask for more. This way we are enforcing Providers to plan their resources consumption beforehand sending WTCT. 

_Technical note_:
This change rewrites `TaskClient` class and changes API of the base `Task` class by adding WTCT hash and `num_subtasks` attribute to client acceptance methods.

### CODE REVIEW

File: golem/task/taskclient.py [hunk 1]
```diff
@@ -23,51 +24,57 @@ def __getstate__(self):
         return state
 
     @staticmethod
-    def assert_exists(node_id, node_dict):
+    def assert_exists(node_id: str,
```
Comment (kmazurek) [inline golem/task/taskclient.py]: While we're at it, I think this method name is misguiding - `assert` implies that it will raise an exception if the value is not in the dict being checked. I think something like `try_create` would be more accurate.

File: tests/golem/task/test_taskclient.py [hunk 2]
```diff
@@ -10,19 +10,116 @@ def test(self):
         node_id = str(uuid.uuid4())
         node_dict = {}
 
+        # when
```
Comment (jiivan) [inline tests/golem/task/test_taskclient.py:13]: Is it possible to test all those scenarios separately (in separate test methods)?

Comment (tworec) [reply tests/golem/task/test_taskclient.py:13]: it can be split for sure, but I can do it in a week or two from now on. I think it is not a blocker here

File: golem/task/taskmanager.py [hunk 3]
```diff
@@ -386,20 +387,18 @@ def task_needs_computation(self, task_id: str) -> bool:
             return False
         return True
 
-    def get_next_subtask(
-            self, node_id, task_id, estimated_performance, price,
-            max_resource_size, max_memory_size):
+    # noqa pylint: disable=too-many-arguments,too-many-return-statements
+    def get_next_subtask(self,
```
Comment (kmazurek) [inline golem/task/taskmanager.py:392]: Nitpicking: since you're changing the signature here you could add type annotations to all parameters.

File: tests/golem/task/test_taskclient.py (hunk 2 above)
Comment (kmazurek) [reply tests/golem/task/test_taskclient.py:13]: I agree with @jiivan here, the test should be split into several separate methods.

**Comment by tworec** [comment]
@jiivan @zakaprov since your review is kinda stylistic, can you approve for now to enable this to be included in release tests? I will address your comments in separate PR next week, ok?

**Review by kmazurek — COMMENTED** [review]
The logic looks okay to me, I left comments about some potential improvements. With that said, I'd recommend running the integration test suite on this branch before merging. Also, it's a large enough change to call for an end-to-end test of its own.

**Review by kmazurek — APPROVED** [review]

File: golem/task/taskclient.py (hunk 1 above)
Comment (tworec) [reply golem/task/taskclient.py]: I've come up with `get_or_initialize`, which is most descriptive of what I can imagine

File: tests/golem/task/test_taskclient.py (hunk 2 above)
Comment (tworec) [reply tests/golem/task/test_taskclient.py:13]: done

Comment (tworec) [reply tests/golem/task/test_taskclient.py:13]: @zakaprov @jiivan can you review once again after changes?

File: tests/golem/task/test_taskmanager.py [hunk 4]
```diff
@@ -330,54 +328,35 @@ def test_get_next_subtask(self, *_):
...
+        assert self.tm.is_my_task("xyz")
+        assert self.tm.get_next_subtask("DEF", "xyz", 1000, 10, None) is None
 
-        wrong_task = not self.tm.is_my_task("xyz")
-        subtask = self.tm.get_next_subtask(
-            "DEF", "xyz", 1000, 10, 5, 10, None)
-        assert subtask is None
-        assert not wrong_task
+        assert self.tm.is_my_task("xyz")
+        assert self.tm.get_next_subtask("DEF", "xyz", 1000, 10, None) is None
```
Comment (tworec) [inline tests/golem/task/test_taskmanager.py]: I'm not sure if those tree checks are adding any value here. Wouldn't be better to replace them with one?

File: apps/core/task/coretask.py [hunk 5]
```diff
@@ -384,8 +379,6 @@ def interpret_task_results(self, subtask_id, task_results, sort=True):
 
     @handle_key_error
     def result_incoming(self, subtask_id):
-        self.counting_nodes[self.subtasks_given[
```
Comment (kmazurek) [inline apps/core/task/coretask.py:387]: Why is this removed?

File: apps/core/task/coretask.py [hunk 6]
```diff
@@ -476,25 +469,27 @@ def _get_resources_root_dir(self):
...
+                      node_id: str,
+                      wtct_hash: Union[None, bytes] = None,
+                      num_subtasks: int = 1
+                     ) -> AcceptClientVerdict:
+        verdict = self.should_accept_client(node_id, wtct_hash)
 
-        if verdict == AcceptClientVerdict.ACCEPTED:
-            client = TaskClient.assert_exists(node_id, self.counting_nodes)
-            client.start()
+        client = TaskClient.get_or_initialize(node_id, self.counting_nodes)
```
Comment (kmazurek) [inline apps/core/task/coretask.py]: This part is always called, regardless of the verdict obtained above. Is that intentional?

File: apps/core/task/coretask.py [hunk 7]
```diff
@@ -476,25 +469,27 @@ def _get_resources_root_dir(self):
...
         return os.path.dirname(prefix)
 
-    def should_accept_client(self, node_id):
-        client = TaskClient.assert_exists(node_id, self.counting_nodes)
-        finishing = client.finishing()
-        max_finishing = self.max_pending_client_results
-
+    def should_accept_client(self,
+                             node_id: str,
+                             wtct_hash: Union[None, bytes] = None
```
Comment (kmazurek) [inline apps/core/task/coretask.py]: I think it's more readable to use `Optional[bytes]` instead of `Union[None, bytes]` here.

File: apps/core/task/coretask.py [hunk 8]
```diff
@@ -476,25 +469,27 @@ def _get_resources_root_dir(self):
...
+        elif client.should_wait(wtct_hash):
             return AcceptClientVerdict.SHOULD_WAIT
 
         return AcceptClientVerdict.ACCEPTED
 
-    def accept_client(self, node_id):
-        verdict = self.should_accept_client(node_id)
+    def accept_client(self,
+                      node_id: str,
+                      wtct_hash: Union[None, bytes] = None,
```
Comment (kmazurek) [inline apps/core/task/coretask.py]: ditto

File: golem/task/taskbase.py [hunk 9]
```diff
@@ -337,11 +337,18 @@ def to_dictionary(self):
         pass
 
     @abc.abstractmethod
-    def should_accept_client(self, node_id):
+    def should_accept_client(self,
+                             node_id: str,
+                             wtct_hash: Union[None, bytes] = None
```
Comment (kmazurek) [inline golem/task/taskbase.py]: Like above, please use `Optional[bytes]` here.

File: golem/task/taskbase.py [hunk 10]
```diff
@@ -337,11 +337,18 @@ def to_dictionary(self):
...
+                             node_id: str,
+                             wtct_hash: Union[None, bytes] = None
+                            ) -> AcceptClientVerdict:
         pass
 
     @abc.abstractmethod
-    def accept_client(self, node_id):
+    def accept_client(self,
+                      node_id: str,
+                      wtct_hash: Union[None, bytes] = None,
```
Comment (kmazurek) [inline golem/task/taskbase.py]: ditto

File: golem/task/taskclient.py [hunk 11]
```diff
@@ -23,51 +24,67 @@ def __getstate__(self):
         return state
 
     @staticmethod
-    def assert_exists(node_id, node_dict):
+    def get_or_initialize(node_id: str,
+                          node_dict: Dict[str, 'TaskClient']) -> 'TaskClient':
+        """ If given `node_id` is already in `node_dict` it's corresponding
```
Comment (kmazurek) [inline golem/task/taskclient.py]: nitpicking: it's -> its

File: golem/task/taskmanager.py [hunk 12]
```diff
@@ -387,27 +388,22 @@ def task_needs_computation(self, task_id: str) -> bool:
...
-    def get_next_subtask(
-            self, node_id, task_id, estimated_performance, price,
-            max_resource_size, max_memory_size):
+    # noqa pylint: disable=too-many-arguments,too-many-return-statements
+    def get_next_subtask(self,
+                         node_id: str,
+                         task_id: str,
+                         estimated_performance: float,
+                         price: int,
+                         wtct_hash,) -> Optional[ComputeTaskDef]:
```
Comment (kmazurek) [inline golem/task/taskmanager.py]: nit: `wtct_hash` has no type hint.

File: golem/task/tasksession.py [hunk 13]
```diff
@@ -408,92 +398,96 @@ def _offer_chosen(
...
-            price=price,
-            size=package_size,
-            resources_options=self.task_server.get_share_options(
-                ctd['subtask_id'], self.address).__dict__
-        )
-        ttc.generate_ethsig(self.my_private_key)
-        if ttc.concent_enabled:
+        wtct_hash = msg.get_short_hash()
+        for _i in range(msg.num_subtasks):
+            ctd = self.task_manager.get_next_subtask(
```
Comment (kmazurek) [inline golem/task/tasksession.py:415]: As discussed offline - the entire body of this loop could be moved to a separate method of `TaskSession`, e.g. `_assign_new_subtask(self, msg: message.tasks.WantToComputeTask, wtct_hash: bytes)`.

File: tests/golem/task/test_taskmanager.py (hunk 4 above)
Comment (kmazurek) [reply tests/golem/task/test_taskmanager.py]: I agree, no point in keeping it this way.

**Review by kmazurek — COMMENTED** [review]
Btw, are there any plans for adding integration tests of this feature?

File: apps/core/task/coretask.py (hunk 6 above)
Comment (tworec) [reply apps/core/task/coretask.py]: good catch, thx.

File: apps/core/task/coretask.py (hunk 5 above)
Comment (tworec) [reply apps/core/task/coretask.py:387]: bc there is no `finish` anymore in taskclient. I've added description for this class since it was rewritten in large part here.

File: tests/golem/task/test_taskmanager.py (hunk 4 above)
Comment (tworec) [reply tests/golem/task/test_taskmanager.py]: removed

File: apps/core/task/coretask.py (hunk 7 above)
Comment (tworec) [reply apps/core/task/coretask.py]: done

File: golem/task/taskclient.py (hunk 11 above)
Comment (tworec) [reply golem/task/taskclient.py]: done

File: golem/task/taskmanager.py (hunk 3 above)
Comment (tworec) [reply golem/task/taskmanager.py:392]: done

File: golem/task/taskmanager.py (hunk 12 above)
Comment (tworec) [reply golem/task/taskmanager.py]: done

**Review by tworec — COMMENTED** [review]
@zakaprov pls look again

File: golem/task/tasksession.py (hunk 13 above)
Comment (tworec) [reply golem/task/tasksession.py:415]: this is not so straightforward when there are `return` and `yield` statements. Lets keep things simple.

**Comment by tworec** [comment]
> Btw, are there any plans for adding integration tests of this feature?

would love to do, but don't have knowledge... who can help?

**Comment by kmazurek** [comment]
> > Btw, are there any plans for adding integration tests of this feature?
> 
> would love to do, but don't have knowledge... who can help?

As discussed offline - @tworec is going to check whether it's possible to cover this use-case using Golem node integration tests (without the involvement of a Golem Unlimited gateway). Let me know should you need more support with writing these tests.

**Review by kmazurek — APPROVED** [review]
LGTM
lgtm
